### PR TITLE
Enable service discovery

### LINKED ISSUE #121: Document service discovery
As a user of services deployed by this operator I'd like to know how to discover its connection details.

It's done when

- documentation is available on all the objects (e.g. ConfigMaps) created by this operator and the circumstances under which they are created and
- documentation on the contents of the created objects is available.

NOTE: This ticket is part of an epic and autocreated for all our operators. It _might_ not apply to this operator in particular, in that case please comment and close
https://github.com/stackabletech/documentation/issues/86

### LINKED ISSUE #137: Enable service discovery
As a user of HDFS I'd like to be able to discover all the information I need to connect to HDFS from a ConfigMap.

This task is done when the HDFS operator:
- [x] writes a `ConfigMap` with the discovery information for name nodes
- [x] the discovery information should be a complete and valid hdfs-site.xml + core-site.xml with all the minimal information needed to connect to HDFS
- [x] optional: Have another (similar to the other operators) fiel dthat just exposes the namenode(s)

### PR DESCRIPTION
**Pull request by maltesander**
## Description

- added config builder for `hdfs-site.xml` and `core-site.xml` to reuse code for the rolegroup and discovery configmaps.
- exposes a configmap with essential properties for clients to connect to this hdfs cluster
- added service discovery docs

HBase PR thats works with this https://github.com/stackabletech/hbase-operator/pull/153

closes https://github.com/stackabletech/hdfs-operator/issues/137
closes https://github.com/stackabletech/hdfs-operator/issues/121

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by maltesander** [comment]
There is a little more work in terms of overrides to reflect that in the discovery.

**Comment by soenkeliebau** [comment]
Are the changes to the product config intentional in here?

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -416,11 +416,18 @@ impl Configuration for NameNodeConfig {
...
         &self,
-        _resource: &Self::Configurable,
+        resource: &Self::Configurable,
         _role_name: &str,
-        _file: &str,
+        file: &str,
     ) -> Result<BTreeMap<String, Option<String>>, ConfigError> {
-        Ok(BTreeMap::new())
+        let mut config = BTreeMap::new();
+        if file == HDFS_SITE_XML {
```
Comment (soenkeliebau) [inline rust/crd/src/lib.rs:424]: This seems unrelated to the discovery?

File: rust/operator/src/hdfs_controller.rs [hunk 2]
```diff
@@ -72,6 +74,21 @@ pub async fn reconcile_hdfs(
     let namenode_podrefs = hdfs.pod_refs(&HdfsRole::NameNode)?;
     let journalnode_podrefs = hdfs.pod_refs(&HdfsRole::JournalNode)?;
 
+    let discovery_cm = build_discovery_configmap(&hdfs, &namenode_podrefs).map_err(|e| {
```
Comment (soenkeliebau) [inline rust/operator/src/hdfs_controller.rs:77]: Maybe a stupid question, but isn't this written too early? If we write the configmap here and roll out new nodes in a later step (for example when first building the cluster) wouldn't those be missing from the configmap?

File: rust/operator/src/hdfs_controller.rs [hunk 3]
```diff
@@ -187,132 +204,47 @@ fn rolegroup_config_map(
...
-        ),
-        (
-            String::from("ha.zookeeper.quorum"),
-            "${env.ZOOKEEPER}".to_string(),
-        ),
-    ]);
+    let mut hdfs_site_xml = String::new();
+    let mut core_site_xml = String::new();
+
+    for (property_name_kind, config) in rolegroup_config {
```
Comment (soenkeliebau) [inline rust/operator/src/hdfs_controller.rs:210]: These changes also seem unrelated to the discovery?

File: rust/operator/src/lib.rs [hunk 4]
```diff
@@ -1,3 +1,6 @@
+//mod discovery;
```
Comment (soenkeliebau) [inline rust/operator/src/lib.rs]: leftover?

File: rust/operator/src/hdfs_controller.rs (hunk 3 above)
Comment (maltesander) [reply rust/operator/src/hdfs_controller.rs:210]: well i replaced the hardcoded config part with that builder thing to reuse it in the discovery, so it kinda is :)

File: rust/operator/src/lib.rs (hunk 4 above)
Comment (maltesander) [reply rust/operator/src/lib.rs]: removed

File: rust/crd/src/lib.rs (hunk 1 above)
Comment (maltesander) [reply rust/crd/src/lib.rs:424]: Yes but required for the product config part.

File: rust/operator/src/hdfs_controller.rs (hunk 2 above)
Comment (maltesander) [reply rust/operator/src/hdfs_controller.rs:77]: The namenode_podrefs include all namenodes that will be created. Its calculated HdfsCluster::pod_refs().

**Comment by maltesander** [comment]
> Are the changes to the product config intentional in here?

Yes i changed some hardcoded stuff here and there and had to remove some defaults and if it is required or not.
